### ISSUE TITLE
Enhancement: [switch-exhaustiveness-check] Support for enums

### ISSUE THREAD

**Issue by SimonSimCity** [typescript-eslint/typescript-eslint#7134]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+plugin+rule+option%22) and found none that match my proposal.
- [X] I have searched the [current rule list](https://typescript-eslint.io/rules/#supported-rules) and found no rules that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [X] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Link to the rule's documentation

https://typescript-eslint.io/rules/switch-exhaustiveness-check/

### Description

As of now, the documentation states that this rule only works on union types. I'd like to extend this to also support enums.

### Fail

```typescript
enum Direction {
  Up,
  Down,
  Left,
  Right,
}

const myVar = Direction.Up;

switch(myVar) {
  case Direction.Up:
    break;
}
```

### Pass

```typescript
enum Direction {
  Up,
  Down,
  Left,
  Right,
}

const myVar = Direction.Up;

switch(myVar) {
  case Direction.Up:
    break;
  case Direction.Down:
    break;
  case Direction.Left:
    break;
  case Direction.Right:
    break;
}

/// OR

enum Direction {
  Up,
  Down,
  Left,
  Right,
}

const myVar = Direction.Up;

switch(myVar) {
  case Direction.Up:
    break;
  default:
    break;
}
```

### Additional Info

Since enums (as far as the compiler reaches) will only have the states which are defined on the enum, a programmer should be able to trust that only one of those values will be assigned to the variable. Assuming this, it's not necessary to provide a `default` option, when all the options are listed as cases.

**Comment by bradzacher** [comment]
So there's a funny thing here in that your examples don't error because TS is being super refined.

Your assumption is that `const val = Direction.Up` creates a variable of type `Direction`, [however it actually creates a variable of type `Direction.Up`](https://www.typescriptlang.org/play?#code/KYOwrgtgBAIglgJ2AYwC5wPYigbwFBRQCqADgDQGwYDuIFhAMsAGar1QBKcA5gBZt4AvnjzIsAZ1RQIATwBqAQwRQAvLEQp0WAHSkA3HgD0hwqYB6AfjzjqcVMl4AKWYoQBKXJWQLxwdUjRMEF0SAC5KQgAjJAUAawNhIA)

So because of this your switch is actually exhaustive because there is just one value in the union!

Even manually annotating the variable doesn't do exactly what you want because TS is annoyingly smart.

Instead if we `declare const val: Direction` or wrap the switch in a function with a parameter typed as `Direction`, then it won't be a single member and will be the entire enum.

[playground][https://typescript-eslint.io/play/#ts=4.5.5&jsx=true&sourceType=module&code=KYOwrgtgBAIglgJ2AYwC5wPYigbwFBRQCqADgDQGwYDuIFhAMsAGar1QBKcA5gBZt4AvnjwATFABsAhkijIsAZ1RQIATwBqMgFyxEKdFgDcIhdTipkvKAAo1mhAEpclZFIXBdSNJhAA6UlqUhABGSFIA1sbCQA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6ZAd0vzIAt6AHrwCGsZPkoA3RExTJafRGQDW6SImjQA9tEjgwAXxAGgA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA)

---

Fun fact - enums are just secret union types! Under the hood an enum is both a type `type Direction = Direction.Up | Direction.Down | ...` and a value which is essentially an object.

**Comment by SimonSimCity** [comment]
@bradzacher Thanks a lot for the time and effort for research and explaining! Just found out that I totally mixed up my configuration with the rules `switch-exhaustiveness-check` and `default-case` and was confused on this when writing this post 🙈
